### ISSUE TITLE
In a mixed enviroment of chds passed as 'possible parents' you probably don't want to fail if some of them are not candidates

### ISSUE THREAD

**Issue by i30817** [Manorhos/imageparse#3]
https://github.com/Manorhos/imageparse/blob/b95a70560d24fff17c1cdc01efb5cee91f070e20/src/chd.rs#L149

Basically just ignore them, instead of sending a error here.

~~In fact the same error sent above for the leaf chd, is probably unnecessary too since you just tested it had a parent, so the parent_sha1 exists. Therefore the error has no need to exist right? Just the ParentsNotFound is enough.~~

edit: i kind of realized this second paragraph is wrong, if you don't want to potentially mix V3 and V4 and V5 types. Never mind. Still the first seems right, you do want to ignore the parent chds that don't have a v5 version as candidates instead of sending a exception i think. A continue should do.

**Comment by i30817** [comment]
If you want backwards compatibility and a smaller api surface you could check if the child and possible parents have all the same version instead of sending a error.

Then 'has_parents' would take care of filtering those versions without parent capability and there would be no risk of mixing versions.

**Comment by Manorhos** [comment]
I considered just skipping CHDs which wouldn't open due to an error or with a version mismatch but decided against it because I thought it would be quite frustrating to pass a parent you know is the correct one just to get back the `ParentNotFound` error.

Also I limited it to V5 because I don't know which implications accepting the older versions has, frankly I haven't seen any pre-V5 CD-ROM CHDs yet.
Would it be valid to mix them?
V1 and V2 also seem to have some sort of parent-child support, but matching is done with MD5s instead of SHA1s.
V3 still has parent_md5, but V4 dropped it, so does that mean V3 is cross-compatible with V1 and V2?

Maybe I can just accept the older versions and forbid mixing for now, but I have no way of testing it. I'd need an old version of chdman to create pre-V5 CHDs, right?

**Comment by i30817** [comment]
> I considered just skipping CHDs which wouldn't open due to an error or with a version mismatch but decided against it because I thought it would be quite frustrating to pass a parent you know is the correct one just to get back the `ParentNotFound` error.

It's even more frustrating for a application to scan a directory for 'possible parents' pass them to the function that opens with parents if the chd it wants to open just happens to have parents, then fails because one of the unrelated chds on the directory was v4 instead of v5.

If you want to prevent mixing versions on the parent child relationship (i agree with this precaution btw), you can pass the version of the main file and only consider those chds that match the version, instead of sending a error that interrupts everything and leaves the calling program with no way to 'fix' it without reading the headers itself.

Not all programs are like mame, that knows what files are the parents from external metadata software lists, so this factory method is supposed to work for both usecases; where you both know or not know if the files you pass are parents, and verifies it when loading and ignores the rest.

~~I don't think supporting md5 parent versions is good idea.~~
Eh seems to harm no one, if you don't mix versions.

**Comment by i30817** [comment]
I was wrong chd-rs has a different method

https://github.com/SnowflakePowered/chd-rs/blob/1e09072d60a6874fc8169addf47eecfb374c5636/chd-rs/src/header.rs#L411

But:
https://github.com/SnowflakePowered/chd-rs/blob/1e09072d60a6874fc8169addf47eecfb374c5636/chd-rs/src/header.rs#L443

and:
https://github.com/SnowflakePowered/chd-rs/blob/1e09072d60a6874fc8169addf47eecfb374c5636/chd-rs/src/header.rs#L472

so basically, V4-V5 seem sane because they all have the parent_sha1 if the has_parent(). But V1-V3 might get a 'md5 parent' (and v3 in particular might get both or either, not sure).

I think this can be done by checking md5 in V1-V2, sha1 in V3-V5 and only allowing comparing the same version as the 'leaf' file, and skip the others.

**Comment by Manorhos** [comment]
Okay, done. Now all CHDs with a `parent_sha1` and/or a `sha1` are matched. I'm not doing the version check though because in my eyes it seems like they should be compatible, and if not, chd-rs should probably complain about them not being compatible.
